p2p/p2p-core/src/ban.rs

```diff
@@ -0,0 +1,19 @@
+//! Data structures related to bans.
+
+use std::time::{Duration, Instant};
+
+use crate::NetZoneAddress;
+
+/// Data within [`crate::services::AddressBookRequest::SetBan`].
+pub struct SetBan<A: NetZoneAddress> {
+    pub address: A,
+    pub ban: bool,
+    pub duration: Duration,
+}
+
+/// Data within [`crate::services::AddressBookResponse::GetBans`].
+pub struct BanState<A: NetZoneAddress> {
+    pub address: A,
+    pub banned: bool,
+    pub unban_instant: Instant,
+}
```

Comment:
### Types (address book)
Some of the requests/responses in this PR require defining some types.

A lot of these are 1-1 with the RPC types but I decided on making more specific versions because:
- Pulling in `cuprate-rpc-types` everywhere seems wrong
- The RPC types are JSON-like, so a `[u8; 32]` field will be something like `String`, adding these types lets us use the canonical type internally

Some of the types are defined in the crate where the request/response is defined, many are defined in `cuprate-types`.

Reply:
Yeah separating them is a good idea IMO

Reply:
Those types seem weird though, wouldn't this make more sense: 

```suggestion
//! Data structures related to bans.

use std::time::{Duration, Instant};

use crate::NetZoneAddress;

/// Data within [`crate::services::AddressBookRequest::SetBan`].
pub struct SetBan<A: NetZoneAddress> {
    pub address: A,
    pub ban: Option<Duration>,
}

/// Data within [`crate::services::AddressBookResponse::GetBans`].
pub struct BanState<A: NetZoneAddress> {
    pub address: A,
    pub banned: Option<Instant>,
}
```